### PR TITLE
feat(data-classes): decorator to instantiate data_classes and docs updates

### PR DESCRIPTION
**Pull request by michaelbrewer**
**Issue #, if available:**
#434
https://github.com/awslabs/aws-lambda-powertools-roadmap/issues/19

## Description of changes:

Changes:
- feat: Create a convenient decorator to instantiate `data_classes` instances
- docs: Update docs to include `@event_source` example
- docs: Add missing docs for `ALBEvent` and format examples to work with PyCharm
- fix: Fix `_generate_hash` to handle the `data` being an instance of `DictWrapper`

```
TypeError: Object of type APIGatewayProxyEventV2 is not JSON serializable
```

### Example usage

```python3
from typing import Dict, Any

from aws_lambda_powertools.utilities.data_classes import APIGatewayProxyEventV2, event_source
from aws_lambda_powertools.utilities.typing import LambdaContext


@event_source(data_class=APIGatewayProxyEventV2)
def lambda_handler(event: APIGatewayProxyEventV2, context: LambdaContext) -> Dict[str, Any]:
    assert event.get_header_value("x-foo") == "Foo"
```

### And can be used with any of the existing data classes like:

```python3
from aws_lambda_powertools.utilities.data_classes import event_source
from aws_lambda_powertools.utilities.data_classes.connect_contact_flow_event import (
    ConnectContactFlowChannel,
    ConnectContactFlowEndpointType,
    ConnectContactFlowEvent,
    ConnectContactFlowInitiationMethod,
)


@event_source(data_class=ConnectContactFlowEvent)
def lambda_handler(event: ConnectContactFlowEvent, context):
    assert event.contact_data.attributes == {"Language": "en-US"}
    assert event.contact_data.channel == ConnectContactFlowChannel.VOICE
    assert event.contact_data.customer_endpoint.endpoint_type == ConnectContactFlowEndpointType.TELEPHONE_NUMBER
    assert event.contact_data.initiation_method == ConnectContactFlowInitiationMethod.API
```

### And this still works with the existing idempotent handler

```python3
persistence_layer = DynamoDBPersistenceLayer(table_name="IdempotencyTable")

@event_source(data_class=APIGatewayProxyEventV2)
@idempotent(persistence_store=persistence_layer)
def lambda_handler(event: APIGatewayProxyEventV2, context):
    assert isinstance(event, APIGatewayProxyEventV2)
    ...
```

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by michaelbrewer** [comment]
Should i update the docs then to include this?

**Comment by heitorlessa** [comment]
Confirming here that I have blocked time to review and get this incorporated tomorrow.

Just been abnormally busy in the last few days with a customer project

File: tests/functional/idempotency/test_idempotency.py [hunk 1]
```diff
@@ -845,3 +848,41 @@ def handler(event, context):
...
+        assert data_record.idempotency_key == self.expected_idempotency_key
+
+    def _update_record(self, data_record: DataRecord) -> None:
+        assert data_record.idempotency_key == self.expected_idempotency_key
+
+    def _get_record(self, idempotency_key) -> DataRecord:
+        ...
+
+    def _delete_record(self, data_record: DataRecord) -> None:
+        ...
```
Comment (heitorlessa) [inline tests/functional/idempotency/test_idempotency.py:868]: I'm on the fence about this one - I might remove it in the future if we end up not reusing it elsewhere in the tests.

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py [hunk 2]
```diff
@@ -224,6 +224,8 @@ def _generate_hash(self, data: Any) -> str:
             Hashed representation of the provided data
 
         """
+        if hasattr(data, "raw_event"):
+            data = data.raw_event
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/persistence/base.py:228]: `hasattr` doesn't work in the way you might think it does - it's a misnomer in Python, really. It'll effectively do a blanket try/except, throw the results away, and specially in this case it'll access all getters from `data` -- the more getters the worse the performance.

Here's a safer alternative which also saves us a branching:

```suggestion
data = getattr(data, "raw_event", data)  # could be a data class depending on decorator order
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
One minor change to make it safer and more performant for the future, all rest LGTM :)

I don't have permission to push to your fork so will push the next release to early next week to ensure we incorporate these in time

File: tests/functional/idempotency/test_idempotency.py (hunk 1 above)
Comment (michaelbrewer) [reply tests/functional/idempotency/test_idempotency.py:868]: We could use it to verify if the parser works with idempotent decorator.

**Comment by michaelbrewer** [comment]
@heitorlessa  does this look go to you now?

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
Yes, thanks a lot Mike!

**Pull request merged**